### ISSUE TITLE
not receiving extra attributes

### ISSUE THREAD

**Issue by pbunyasrie** [biola/rack-cas#36]
I'm on rails 4.2.1, and am I'm able to login fine using CAS. However, I'm getting an empty hash when trying to view extra_attributes.

I've already whitelisted the attributes in config/application.rb like so:
`config.rack_cas.extra_attributes_filter = %w(groupMembership mail)`

I'm using Jasig CAS 3.5.3, and defining the attributes that the service can see in  the Services Management Console as decribed in https://wiki.jasig.org/display/CASUM/Services+Management

At this point, I'm using pretty much the same same controller and view code as the demo.

Any ideas?

**Comment by adamcrown** [comment]
Hmm, I wonder if they changed the way they send extra attributes. I don't currently have a Jasig CAS server to test against. Would you mind doing some legwork so we can figure out what the XML response looks like?

- Clone `git@github.com:biola/rack-cas.git`
- Replace the `gem 'rack-cas'` line in your `Gemfile` with `gem 'rack-cas', path: '../rack-cas'`
- Add `gem 'pry'` to your `Gemfile`
- Do a `bundle install`
- Restart your server
- Add `require 'pry'; binding.pry` to the top of the `extra_attributes` method in `lib/rack-cas/service_validation_response.rb`
- Call `response.body` in the pry session and send me the result.

Thanks.

**Comment by pbunyasrie** [comment]
[1] pry(#<RackCAS::ServiceValidationResponse>)> response.body
    => "\r\n\r\n\r\n\r\n<cas:serviceResponse xmlns:cas='http://www.yale.edu/tp/cas' xmlns:vcu='http://login.vcu.edu/cas'>\r\n\t<cas:authenticationSuccess>\r\n\t\t<cas:user>test-studnt2</cas:user>\r\n\t\t\r\n\t\t\r\n\t\t<vcu:attributes>\r\n\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\t<vcu:sn>Rolled</vcu:sn>\r\n\t\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\t\r\n\t\t\t\t\t\t\t<vcu:givenName>En</vcu:givenName>\r\n\t\t\t\t\t\t\r\n\t\t\t\t\t\t\t<vcu:givenName>En 2</vcu:givenName>\r\n\t\t\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\t\r\n\r\n\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\r\n\t\t\t\t\t\t<vcu:displayName>En 2 Rolled</vcu:displayName>\r\n\t\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t\r\n\t\t</vcu:attributes>\r\n\t</cas:authenticationSuccess>\r\n</cas:serviceResponse>\r\n\r\n"
    [2] pry(#<RackCAS::ServiceValidationResponse>)> exit

**Comment by adamcrown** [comment]
So cleaning things up a bit your response looks like this

```xml
<cas:serviceResponse xmlns:cas='http://www.yale.edu/tp/cas' xmlns:vcu='http://login.vcu.edu/cas'>
  <cas:authenticationSuccess>
    <cas:user>test-studnt2</cas:user>
    <vcu:attributes>
            <vcu:sn>Rolled</vcu:sn>
              <vcu:givenName>En</vcu:givenName>
              <vcu:givenName>En 2</vcu:givenName>
            <vcu:displayName>En 2 Rolled</vcu:displayName>
    </vcu:attributes>
  </cas:authenticationSuccess>
</cas:serviceResponse>
```

rack-cas expects all those `<vcu:...>` tags to be `<cas:...>`. Do you know why they're different in your case. If that's a common thing and not some sort of hack then we can add some configuration to support that.

**Comment by pbunyasrie** [comment]
Thanks for looking into this. The maintainer of our CAS system said he does not know why the namespace is different - it was implemented by the person before him, who is no longer here.

I've confirmed that rack-cas can successfully return attributes for my XML response by changing:

      # Jasig style
      if attr_node = xml.at('/cas:serviceResponse/cas:authenticationSuccess/cas:attributes')

to

      # Jasig style
      if attr_node = xml.at('/cas:serviceResponse/cas:authenticationSuccess/vcu:attributes')

in service_validation_response.rb

Not really sure how to make this configurable though...or if it's even a common situation. It would be nice if it was configurable somehow.

Using other CAS clients such as phpCAS, the attributes are returned correctly for our namespace.

**Comment by adamcrown** [comment]
I've marked this issue as an enhancement. I'll see if I can get a solution into the code, but I can't promise I'll be able to get to it soon. For now, I'd say your best bet is to fork the repo and just make the edits you mentioned above.

**Comment by dsusco** [comment]
Just posted this: https://github.com/biola/rack-cas/pull/54

**Comment by adamcrown** [comment]
This issue should be resolved now with #54
